sdk/typescript/src/cryptography/keypair.ts

```diff
@@ -15,26 +16,51 @@ export type ExportedKeypair = {
 	privateKey: string;
 };
 
+interface SignedMessage {
+	bytes: Uint8Array;
+	signature: Uint8Array;
+}
+
 /**
- * A keypair used for signing transactions.
+ * TODO: Document
  */
-export interface Keypair {
+export abstract class Keypair {
+	abstract sign(bytes: Uint8Array): Promise<Uint8Array>;
+
+	async signWithIntent(bytes: Uint8Array, intent: IntentScope): Promise<SignedMessage> {
+		const intentMessage = messageWithIntent(intent, bytes);
+		const signature = await this.sign(intentMessage);
+
+		return {
+			bytes: intentMessage,
+			signature,
+		};
+	}
+
+	async signTransactionBlock(bytes: Uint8Array) {
+		return this.signWithIntent(bytes, IntentScope.TransactionData);
+	}
+
+	async signMessage(bytes: Uint8Array) {
+		return this.signWithIntent(bytes, IntentScope.PersonalMessage);
+	}
+
 	/**
 	 * The public key for this keypair
 	 */
-	getPublicKey(): PublicKey;
+	abstract getPublicKey(): PublicKey;
```

Comment:
Maybe a stupid question, but do all keypairs have a public key?  (Eg, you mentioned multisig would be a keypair)

Reply:
Yeah good question, I think the answer is probably yes, but we can confirm. But some of them dont have accessible private keys, so things like export aren't possible. So we will need to rework a little bit of this eventually.